Fix/add csrf token to js post requests

Simrayz opened the pull request:
## Scope and purpose

Fixes bugs identified in #3396.

Closes #3437, Closes #3438, Closes #3439 and Closes #3441.

## Contributor Checklist

* [ ] Added a changelog fragment for [towncrier](https://nav.readthedocs.io/en/latest/hacking/hacking.html#adding-a-changelog-entry)
* [ ] Added/amended tests for new/changed code
* [ ] Added/changed documentation
* [x] Linted/formatted the code with ruff, easiest by using [pre-commit](https://nav.readthedocs.io/en/latest/hacking/hacking.html#pre-commit-hooks-and-ruff)
* [x] The first line of the commit message continues the sentence "If applied, this commit will ...", starts with a capital letter, does not end with punctuation and is 50 characters or less long. See https://cbea.ms/git-commit/
* [x] This pull request is based on the correct upstream branch: For a patch/bugfix affecting the latest stable version, it should be based on that version's branch (`<major>.<minor>.x`). For a new feature or other additions, it should be based on `master`.
* [ ] If applicable: Created new issues if this PR does not fix the issue completely/there is further work to be done
* [ ] If this results in changes in the UI: Added screenshots of the before and after
* [ ] If this adds a new Python source code file: Added the [boilerplate header](https://nav.readthedocs.io/en/latest/hacking/hacking.html#python-boilerplate-headers) to that file

<!-- Make this a draft PR if the content is subject to change, cannot be merged or if it is for initial feedback -->

Simrayz commented on python/nav/web/templates/status2/status.html:
```diff
@@ -63,8 +63,8 @@
-    <div id="action-panel-revised" class="row">
-
+    <form id="action-panel-revised" class="row" method="POST">
+      {% csrf_token %}
```
The `status2/views.js` code, `this.$el.find('form').serialize()` apparently serializes fields in all forms on the page. So changed it to `this.$el.find('#status-panel form').serialize()`, and it doesn't do this anymore :)

Simrayz commented on python/nav/web/static/js/src/plugins/navlet_controller.js:
```diff
@@ -198,7 +198,16 @@ define(['libs/urijs/URI', 'libs/spin.min'], function (URI, Spinner) {
-                    var request = $.post(that.removeUrl, {'navletid': that.navlet.id});
+                    const csrfToken = $('#navlets-action-form input[name="csrfmiddlewaretoken"]').val();
+                    const request = $.ajax({
+                       url: this.removeUrl,
```
I did not even register that the code was using "this" and "that", but I changed it back to `that.remove.url`. What a weird way to refer to things instead of using references with separable names.

Simrayz commented on changelog.d/3465.security.md:
```diff
@@ -0,0 +1 @@
+Add CSRF token to ajax post requests
```
Since the only change was to make sonarqube happy (var -> const), I just piggybacked the changes to save time on rebase editing. However, `Add changelog entry` now only contains the new changelog file.

johannaengland commented on python/nav/web/static/js/src/webfront.js:
```diff
@@ -222,7 +228,10 @@ require([
             event.preventDefault();
             feedback.removeAlertbox();
-            var request = $.post(this.getAttribute('action'));
+            const request = $.post(
```
Why are you not sending a CSRF token here?

Simrayz replied:
What is happening here is that some jQuery magic is serializing the form, i.e. extracting the values and converting it to a request payload (with key=value pairs). As such, the code on line 233 `$(this).serialize()` makes sure that the CSRF token is sent with the request. 

This is how it works:
1. jQuery creates a reference to the form by using its ID, ie. `#form-set-default-dashboard`
2. On line 228 the `submit` event is called on the form, which triggers a submission.
3. Within the function callback, `this` refers to the form that just submitted.
4. As `this` contains the form values, it can be serialized to automatically include the `csrfmiddlewaretoken`.

johannaengland reviewed (commented):
Mostly some questions. Do you have any reference to how you fixed this? So that I can try understanding this better and if something like this comes up again in the future

Simrayz commented on python/nav/web/templates/neighbors/base.html:
```diff
@@ -2,65 +2,67 @@
+    {% csrf_token %}
+    <div>
+        <ul id="action-buttons" class="button-group">
+            {{ csrf_input }}
```
Good catch, I think this was just me hallucinating the the function that outputs the csrf input is logically named `csrf_input`. I'll remove it!

Simrayz commented:
@johannaengland I have rebased the PR and done the following changes:
* Removed unrelated formatting from commits
* Changed commit messages to include issue number on a second line

I added documentation to Javascript hacking about how to include CSRF tokens in the template, and how to access them for usage in AJAX requests.

In addition I have fixed the 400 bug in `remove-user-navlet`, apparently the django view expects form data and not JSON.

However, I was not able to reproduce the 403 errors you reported. Are you logged in with a user that does not have all required permissions? I usually test with an admin user, or a new user with minimal permissions. I tried creating a new user (not part of any groups), and it was able to Acknowledge alerts.

johannaengland commented on python/nav/web/templates/neighbors/base.html:
```diff
@@ -28,10 +28,10 @@
     <ul id="action-buttons" class="button-group">
       <li>
-        <button id="ignore-selected" class="tiny">Ignore selected</button>
+        <button id="ignore-selected" class="tiny" type="button">Ignore selected</button>
```
Now I'm curious, how did that fix that bug?

johannaengland commented on doc/hacking/javascript.rst:
```diff
@@ -100,3 +100,125 @@ only be global configuration files for *RequireJS*, *jshint*, etc.
+
+**Multiple forms on the same page:**
+
+When you have multiple forms that perform different actions, each form needs its own CSRF token. This example shows two example forms for resoruce operations - one for renaming and one for deleting.
```
```suggestion
When you have multiple forms that perform different actions, each form needs its own CSRF token. This example shows two example forms for resource operations - one for renaming and one for deleting.
```

johannaengland reviewed (approved):
Just a tiny typo fix
